### ISSUE TITLE
Study ways to make a pip intall for Linux (or another mothod to install the program)

### ISSUE THREAD

**Issue by mcjpedro** [mrdrzit/Behavython#26]

**Comment by mrdrzit** [comment]
This could be a light in a dark.. dark forest. 

https://ch-sa.github.io/2022/01/22/pyqt-packaging.html

**Comment by mrdrzit** [comment]
Main problem is the .ui file calling scripts as header files
It gives out this problem

> (PostRock) C:\Users\uzuna\Desktop\Behavython>python
Python 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:18:12) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import Behavython
>>> from Behavython import behavython_front
>>> behavython_front.main()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\uzuna\Desktop\Behavython\Behavython\behavython_front.py", line 137, in main
    window = behavython_gui()                # Create an instance of our class
  File "C:\Users\uzuna\Desktop\Behavython\Behavython\behavython_front.py", line 51, in __init__
    uic.loadUi(pkg_resources.resource_filename(
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\__init__.py", line 238, in loadUi
    return DynamicUILoader(package).loadUi(uifile, baseinstance, resource_suffix)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\Loader\loader.py", line 66, in loadUi
    return self.parse(filename, resource_suffix)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 1037, in parse
    actor(elem)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 828, in createUserInterface
    self.traverseWidgetTree(elem)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 806, in traverseWidgetTree
    handler(self, child)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 273, in createWidget
    self.traverseWidgetTree(elem)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 806, in traverseWidgetTree
    handler(self, child)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 503, in createLayout
    self.traverseWidgetTree(elem)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 806, in traverseWidgetTree
    handler(self, child)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 546, in handleItem
    self.traverseWidgetTree(elem)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 806, in traverseWidgetTree
    handler(self, child)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 273, in createWidget
    self.traverseWidgetTree(elem)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 806, in traverseWidgetTree
    handler(self, child)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 273, in createWidget
    self.traverseWidgetTree(elem)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 806, in traverseWidgetTree
    handler(self, child)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 264, in createWidget
    self.stack.push(self.setupObject(widget_class, parent, elem))
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\uiparser.py", line 228, in setupObject
    obj = self.factory.createQObject(clsname, name, args, is_attribute)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\objcreator.py", line 100, in createQObject
    factory = self.findQObjectType(classname)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\objcreator.py", line 125, in findQObjectType
    w = module.search(classname)
  File "D:\Installs\Anaconda\envs\PostRock\lib\site-packages\PyQt5\uic\Loader\qobjectcreator.py", line 113, in search
    module = __import__(mname, {}, {}, (cls,))
ModuleNotFoundError: No module named 'behavython_front'

Removing these lines also removes the problem (but also remove the ability to plot to the main window):

https://github.com/mrdrzit/Behavython/blob/ee8467292f414ae20021db003b78fbc98004af56/Behavython_GUI.ui#L499-L506

Wondering if there is a way to remove this line but also continue being able to modify the main window to plot the graphs..

**Comment by mcjpedro** [comment]
This issue was a very complicated task. But it's solved by this steps:

- Python modules documentation was studied to understand how the sys.path functions works (
the error was because of indexing the directories). In this document a important piece was found: "Accordingly Python documentation: When a module named spam is imported, the interpreter first searches for a built-in module with that name. These module names are listed in sys.builtin_module_names. If not found, it then searches for a file named spam.py in a list of directories given by the variable sys.path. sys.path is initialized from these locations: (1) The directory containing the input script (or the current directory when no file is specified). (2) PYTHONPATH (a list of directory names, with the same syntax as the shell variable PATH). (3) The installation-dependent default (by convention including a site-packages directory, handled by the site module)." 

- To counter the problem a little command "sys.path.append(os.path.dirname(__file__))" indicating the path to behavython plot class file was add and the problem desapear. But another problem becomes: No module found behavython_back

- In real the problem is the same (bad indexing for directories), so to solve this issue I rebuilded the setup.cfg file using pyproject.toml build and tried to understand the mechanism behind this feature.

- After many attempts I found the solution: instead of call behavython_back, should be called behavython.behavython_back -> THAT MAKES SENSE, because the structure is "package.module", in other words: call packages and after that call the module.

- AMEM
